astropy/io/fits/hdu/hdulist.py

```diff
@@ -1330,6 +1325,22 @@ def _read_next_hdu(self):
                     hdu = _BaseHDU.fromstring(data, **kwargs)
                     self._data = data[hdu._data_offset + hdu._data_size :]
 
+                if not kwargs.get("disable_image_compression", False):
+                    from astropy.io.fits import BinTableHDU, CompImageHDU
+
+                    if isinstance(hdu, BinTableHDU) and CompImageHDU.match_header(
+                        hdu.header
+                    ):
+                        kwargs_comp = {
+                            key: kwargs[key]
+                            for key in {
+                                "scale_back",
+                                "uint",
+                                "do_not_scale_image_data",
+                            }.intersection(kwargs)
+                        }
+                        hdu = CompImageHDU(bintable=hdu, **kwargs_comp)
```

Comment:
I couldn't get this to work before, but will give it another shot

Reply:
I'll check in more detail later but I think the issue was related to the order in which

```
        for c in reversed(list(itersubclasses(cls))):
```

Is executed, and the fact that there is no way to change the priority. However, we could always add a ``__priority__`` attrribute to the classes (defaulting to 0) which would be used as an initial key before alphabetical order (in ``itersubclasses``). What do you think?

Reply:
yeah that's what I was thinking ... this subclass iteration is not very flexible, but maybe we can find a way to prevent BinTableHDU from matching compressed HDUs ? `CompImageHDU` matches ZIMAGE=T so maybe we can just modify BinTableHDU's `match_header` to exclude ZIMAGE=T ?

Reply:
Will investigate!

Reply:
The latter approach would require managing to pass information about whether compression is disabled all the way into match_header which will either add kwargs to a few methods or rely on a global variable like before. I think it would be cleaner to simply allow sub-classes to override the priority: https://github.com/astropy/astropy/pull/16434

Reply:
Hmm I need to check if that fully solves the issue here, investigating some more

Reply:
Ok so this is turning out to be a lot trickier to get to work - the issue is that all the I/O stuff tries to set things like ``_data_offset`` and so on on ``CompImageHDU`` but this class is not really the one that should be dealing with the I/O - that would be ``BinTableHDU``. Could we keep this as-is for now and I can try and work on this in a follow-up PR?

Reply:
Right, this is indeed making things more complicated :(
Let's forget about this for now then.